sesemi/models/heads/loss.py

```diff
@@ -110,7 +118,49 @@ def forward(
         **kwargs,
     ) -> Tensor:
         inputs_u, targets_u = data[self.input_data]
-        x_unlabeled = backbones[self.input_backbone](inputs_u)
-        output_unlabeled = self.fc_unlabeled(x_unlabeled)
-        loss_u = F.cross_entropy(output_unlabeled, targets_u, reduction="none")
+        x_u = backbones[self.input_backbone](inputs_u)
+        output_u = self.fc_unsupervised(x_u)
+        loss_u = F.cross_entropy(output_u, targets_u, reduction="none")
+        return loss_u
+
+
+class EntropyMinimizationLossHead(LossHead):
+    """The entropy minimization loss head.
+    https://papers.nips.cc/paper/2004/file/96f2b50b5d3613adf9c27049b2a888c7-Paper.pdf
+    """
+
+    def __init__(
+        self,
+        input_data: str,
+        input_backbone: str = "backbone",
+        logger: Optional[LightningLoggerBase] = None,
+    ):
+        """Initializes the loss head.
+
+        Args:
+            input_data: The key used to get the unlabeled input data.
+            input_backbone: The key used to get the backbone for feature extraction.
+            logger: An optional PyTorch Lightning logger.
+        """
+        super().__init__(logger)
+        self.input_data = input_data
+        self.input_backbone = input_backbone
+
+    def build(self, backbones: Dict[str, Backbone], heads: Dict[str, Head], **kwargs):
+        self.fc_supervised = heads["supervised_head"]
+
+    def forward(
+        self,
+        data: Dict[str, Any],
+        backbones: Dict[str, Backbone],
+        heads: Dict[str, Head],
+        features: Dict[str, Any],
+        step: int,
+        **kwargs,
+    ) -> Tensor:
+        inputs_u, _ = data[self.input_data]
+        x_u = backbones[self.input_backbone](inputs_u)
+        output_u = self.fc_supervised(x_u)
+        output_u = F.softmax(output_u, dim=1)
+        loss_u = (-output_u * torch.log(output_u + 1e-5)).sum(1)
```

Comment:
You might consider the use of the `log_softmax` function here:
https://pytorch.org/docs/stable/generated/torch.nn.functional.log_softmax.html